### PR TITLE
Implement Issue #389 : Support for MessagePack

### PR DESCRIPTION
**Pull request by lxhoan**
this implementation uses the messagepack-core only

### CODE REVIEW

**Comment by lxhoan** [comment]
@majst01 , @fmachado 
please help to take a look, thank you very much

**Comment by fmachado** [comment]
@lxhoan nice work here! I'll take a look during later.

**Comment by majst01** [comment]
@lxhoan i need some time as well.

**Comment by fmachado** [comment]
@lxhoan I've nothing against reformatting the code but could you please avoid adding it to the same PR you are creating a new feature or fixing a bug? Thanks!

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 1]
```diff
@@ -99,38 +121,72 @@ public InfluxDBImpl(final String url, final String username, final String passwo
...
-        this.retrofit = new Retrofit.Builder()
-                .baseUrl(url)
-                .client(client.addInterceptor(loggingInterceptor).addInterceptor(gzipRequestInterceptor).build())
-                .addConverterFactory(MoshiConverterFactory.create())
-                .build();
-        this.influxDBService = influxDBService;
-        this.adapter = adapter;
+
+    if (ResponseFormat.MSGPACK.equals(responseFormat)) {
+      String[] versionNumbers = version().split("\\.");
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: This should go into its own method which can be then tested more easily.

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 2]
```diff
@@ -139,17 +195,17 @@ public InfluxDBImpl(final String url, final String username, final String passwo
...
   private InetAddress parseHostAddress(final String url) {
-      HttpUrl httpUrl = HttpUrl.parse(url);
+    HttpUrl httpUrl = HttpUrl.parse(url);
 
-      if (httpUrl == null) {
-          throw new IllegalArgumentException("Unable to parse url: " + url);
-      }
+    if (httpUrl == null) {
+      throw new IllegalArgumentException("Unable to parse url: " + url);
+    }
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: Please revert formatting changes as they make reviewing more difficult.

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 3]
```diff
@@ -212,65 +268,54 @@ public InfluxDB enableBatch(final BatchOptions batchOptions) {
...
-    this.batchProcessor = BatchProcessor
-            .builder(this)
-            .actions(batchOptions.getActions())
-            .exceptionHandler(batchOptions.getExceptionHandler())
-            .interval(batchOptions.getFlushDuration(), batchOptions.getJitterDuration(), TimeUnit.MILLISECONDS)
-            .threadFactory(batchOptions.getThreadFactory())
-            .bufferLimit(batchOptions.getBufferLimit())
-            .consistencyLevel(batchOptions.getConsistency())
-            .build();
+    this.batchProcessor = BatchProcessor.builder(this).actions(batchOptions.getActions())
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: same here, formatting with no changes in behavior.

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 4]
```diff
@@ -212,65 +268,54 @@ public InfluxDB enableBatch(final BatchOptions batchOptions) {
...
+        .threadFactory(batchOptions.getThreadFactory()).bufferLimit(batchOptions.getBufferLimit())
+        .consistencyLevel(batchOptions.getConsistency()).build();
     this.batchEnabled.set(true);
     return this;
   }
 
   @Override
-  public InfluxDB enableBatch(final int actions, final int flushDuration,
-                              final TimeUnit flushDurationTimeUnit) {
+  public InfluxDB enableBatch(final int actions, final int flushDuration, final TimeUnit flushDurationTimeUnit) {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 5]
```diff
@@ -212,65 +268,54 @@ public InfluxDB enableBatch(final BatchOptions batchOptions) {
...
     return this;
   }
 
   @Override
-  public InfluxDB enableBatch(final int actions, final int flushDuration,
-                              final TimeUnit flushDurationTimeUnit, final ThreadFactory threadFactory) {
-    enableBatch(actions, flushDuration, flushDurationTimeUnit, threadFactory, (points, throwable) -> { });
+  public InfluxDB enableBatch(final int actions, final int flushDuration, final TimeUnit flushDurationTimeUnit,
+      final ThreadFactory threadFactory) {
+    enableBatch(actions, flushDuration, flushDurationTimeUnit, threadFactory, (points, throwable) -> {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 6]
```diff
@@ -212,65 +268,54 @@ public InfluxDB enableBatch(final BatchOptions batchOptions) {
...
+    });
     return this;
   }
 
   @Override
   public InfluxDB enableBatch(final int actions, final int flushDuration, final TimeUnit flushDurationTimeUnit,
-                              final ThreadFactory threadFactory,
-                              final BiConsumer<Iterable<Point>, Throwable> exceptionHandler,
-                              final ConsistencyLevel consistency) {
+      final ThreadFactory threadFactory, final BiConsumer<Iterable<Point>, Throwable> exceptionHandler,
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 7]
```diff
@@ -212,65 +268,54 @@ public InfluxDB enableBatch(final BatchOptions batchOptions) {
...
     enableBatch(actions, flushDuration, flushDurationTimeUnit, threadFactory, exceptionHandler)
         .setConsistency(consistency);
     return this;
   }
 
   @Override
   public InfluxDB enableBatch(final int actions, final int flushDuration, final TimeUnit flushDurationTimeUnit,
-                              final ThreadFactory threadFactory,
-                              final BiConsumer<Iterable<Point>, Throwable> exceptionHandler) {
+      final ThreadFactory threadFactory, final BiConsumer<Iterable<Point>, Throwable> exceptionHandler) {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 8]
```diff
@@ -212,65 +268,54 @@ public InfluxDB enableBatch(final BatchOptions batchOptions) {
...
     enableBatch(actions, flushDuration, 0, flushDurationTimeUnit, threadFactory, exceptionHandler);
     return this;
   }
 
   private InfluxDB enableBatch(final int actions, final int flushDuration, final int jitterDuration,
-                               final TimeUnit durationTimeUnit, final ThreadFactory threadFactory,
-                               final BiConsumer<Iterable<Point>, Throwable> exceptionHandler) {
+      final TimeUnit durationTimeUnit, final ThreadFactory threadFactory,
+      final BiConsumer<Iterable<Point>, Throwable> exceptionHandler) {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 9]
```diff
@@ -212,65 +268,54 @@ public InfluxDB enableBatch(final BatchOptions batchOptions) {
...
     }
-    this.batchProcessor = BatchProcessor
-            .builder(this)
-            .actions(actions)
-            .exceptionHandler(exceptionHandler)
-            .interval(flushDuration, jitterDuration, durationTimeUnit)
-            .threadFactory(threadFactory)
-            .consistencyLevel(consistency)
-            .build();
+    this.batchProcessor = BatchProcessor.builder(this).actions(actions).exceptionHandler(exceptionHandler)
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 10]
```diff
@@ -337,8 +382,7 @@ public void write(final String database, final String retentionPolicy, final Poi
       HttpBatchEntry batchEntry = new HttpBatchEntry(point, database, retentionPolicy);
       this.batchProcessor.put(batchEntry);
     } else {
-      BatchPoints batchPoints = BatchPoints.database(database)
-                                           .retentionPolicy(retentionPolicy).build();
+      BatchPoints batchPoints = BatchPoints.database(database).retentionPolicy(retentionPolicy).build();
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 11]
```diff
@@ -365,28 +409,16 @@ public void write(final int udpPort, final Point point) {
...
     RequestBody lineProtocol = RequestBody.create(MEDIA_TYPE_STRING, batchPoints.lineProtocol());
-    execute(this.influxDBService.writePoints(
-        this.username,
-        this.password,
-        batchPoints.getDatabase(),
-        batchPoints.getRetentionPolicy(),
-        TimeUtil.toTimePrecision(batchPoints.getPrecision()),
-        batchPoints.getConsistency().value(),
-        lineProtocol));
+    execute(this.influxDBService.writePoints(this.username, this.password, batchPoints.getDatabase(),
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 12]
```diff
@@ -365,28 +409,16 @@ public void write(final int udpPort, final Point point) {
...
-          final TimeUnit precision, final String records) {
-    execute(this.influxDBService.writePoints(
-        this.username,
-        this.password,
-        database,
-        retentionPolicy,
-        TimeUtil.toTimePrecision(precision),
-        consistency.value(),
-        RequestBody.create(MEDIA_TYPE_STRING, records)));
+      final TimeUnit precision, final String records) {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: Same here, i see changes but they are hard to see.

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 13]
```diff
@@ -401,14 +433,12 @@ public void write(final String database, final String retentionPolicy, final Con
     write(database, retentionPolicy, consistency, TimeUnit.NANOSECONDS, records);
   }
 
-
   @Override
   public void write(final String database, final String retentionPolicy, final ConsistencyLevel consistency,
-          final TimeUnit precision, final List<String> records) {
+      final TimeUnit precision, final List<String> records) {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 14]
```diff
@@ -417,25 +447,25 @@ public void write(final int udpPort, final String records) {
     initialDatagramSocket();
     byte[] bytes = records.getBytes(StandardCharsets.UTF_8);
     try {
-        datagramSocket.send(new DatagramPacket(bytes, bytes.length, hostAddress, udpPort));
+      datagramSocket.send(new DatagramPacket(bytes, bytes.length, hostAddress, udpPort));
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 15]
```diff
@@ -417,25 +447,25 @@ public void write(final int udpPort, final String records) {
...
     if (datagramSocket == null) {
-        synchronized (InfluxDBImpl.class) {
-            if (datagramSocket == null) {
-                try {
-                    datagramSocket = new DatagramSocket();
-                } catch (SocketException e) {
-                    throw new InfluxDBIOException(e);
-                }
-            }
+      synchronized (InfluxDBImpl.class) {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 16]
```diff
@@ -476,55 +506,48 @@ public void onFailure(final Call<QueryResult> call, final Throwable throwable) {
    * {@inheritDoc}
    */
   @Override
-    public void query(final Query query, final int chunkSize, final Consumer<QueryResult> consumer) {
+  public void query(final Query query, final int chunkSize, final Consumer<QueryResult> consumer) {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 17]
```diff
@@ -476,55 +506,48 @@ public void onFailure(final Call<QueryResult> call, final Throwable throwable) {
...
-        if (query instanceof BoundParameterQuery) {
-            BoundParameterQuery boundParameterQuery = (BoundParameterQuery) query;
-            call = this.influxDBService.query(this.username, this.password,
-                    query.getDatabase(), query.getCommandWithUrlEncoded(), chunkSize,
-                    boundParameterQuery.getParameterJsonWithUrlEncoded());
-        } else {
-            call = this.influxDBService.query(this.username, this.password,
-                    query.getDatabase(), query.getCommandWithUrlEncoded(), chunkSize);
-        }
+    Call<ResponseBody> call = null;
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 18]
```diff
@@ -534,13 +557,13 @@ public void onFailure(final Call<ResponseBody> call, final Throwable t) {
   public QueryResult query(final Query query, final TimeUnit timeUnit) {
     Call<QueryResult> call = null;
     if (query instanceof BoundParameterQuery) {
-        BoundParameterQuery boundParameterQuery = (BoundParameterQuery) query;
-        call = this.influxDBService.query(this.username, this.password, query.getDatabase(),
-                TimeUtil.toTimePrecision(timeUnit), query.getCommandWithUrlEncoded(),
-                boundParameterQuery.getParameterJsonWithUrlEncoded());
+      BoundParameterQuery boundParameterQuery = (BoundParameterQuery) query;
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 19]
```diff
@@ -563,19 +586,20 @@ public void createDatabase(final String name) {
    */
   @Override
   public void deleteDatabase(final String name) {
-    execute(this.influxDBService.postQuery(this.username, this.password,
-                                           Query.encode("DROP DATABASE \"" + name + "\"")));
+    execute(
+        this.influxDBService.postQuery(this.username, this.password, Query.encode("DROP DATABASE \"" + name + "\"")));
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 20]
```diff
@@ -563,19 +586,20 @@ public void createDatabase(final String name) {
...
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public List<String> describeDatabases() {
-    QueryResult result = execute(this.influxDBService.query(this.username,
-                                                            this.password, SHOW_DATABASE_COMMAND_ENCODED));
+    QueryResult result = execute(
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 21]
```diff
@@ -606,18 +630,17 @@ public boolean databaseExists(final String name) {
   private Call<QueryResult> callQuery(final Query query) {
     Call<QueryResult> call;
     if (query instanceof BoundParameterQuery) {
-        BoundParameterQuery boundParameterQuery = (BoundParameterQuery) query;
-        call = this.influxDBService.postQuery(this.username,
-                this.password, query.getDatabase(), query.getCommandWithUrlEncoded(),
-                boundParameterQuery.getParameterJsonWithUrlEncoded());
+      BoundParameterQuery boundParameterQuery = (BoundParameterQuery) query;
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 22]
```diff
@@ -606,18 +630,17 @@ public boolean databaseExists(final String name) {
...
+          query.getCommandWithUrlEncoded(), boundParameterQuery.getParameterJsonWithUrlEncoded());
     } else {
-        if (query.requiresPost()) {
-          call = this.influxDBService.postQuery(this.username,
-                  this.password, query.getDatabase(), query.getCommandWithUrlEncoded());
-        } else {
-          call = this.influxDBService.query(this.username,
-                  this.password, query.getDatabase(), query.getCommandWithUrlEncoded());
-        }
+      if (query.requiresPost()) {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 23]
```diff
@@ -657,11 +680,11 @@ public void flush() {
   @Override
   public void close() {
     try {
-        this.disableBatch();
+      this.disableBatch();
     } finally {
-        if (datagramSocket != null && !datagramSocket.isClosed()) {
-            datagramSocket.close();
-        }
+      if (datagramSocket != null && !datagramSocket.isClosed()) {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 24]
```diff
@@ -699,13 +722,8 @@ public void createRetentionPolicy(final String rpName, final String database, fi
...
 
     StringBuilder queryBuilder = new StringBuilder("CREATE RETENTION POLICY \"");
-    queryBuilder.append(rpName)
-        .append("\" ON \"")
-        .append(database)
-        .append("\" DURATION ")
-        .append(duration)
-        .append(" REPLICATION ")
-        .append(replicationFactor);
+    queryBuilder.append(rpName).append("\" ON \"").append(database).append("\" DURATION ").append(duration)
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: formatting

File: src/main/java/org/influxdb/msgpack/MessagePackResponseBodyConverter.java [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+ * @author hoan.le [at] bonitoo.io
+ *
+ */
+public class MessagePackResponseBodyConverter implements Converter<ResponseBody, QueryResult> {
+
+  @Override
+  public QueryResult convert(final ResponseBody value) throws IOException {
+      try (InputStream is = value.byteStream()) {
+        MessagePackTraverser traverser = new MessagePackTraverser();
+        for (QueryResult queryResult : traverser.traverse(is)) {
```
[the remark this answers is not in the document]
Comment (lxhoan) [reply src/main/java/org/influxdb/msgpack/MessagePackResponseBodyConverter.java]: In general, there should be only one QueryResult so I return the first one. However this code can look ambiguous. I refactor MessagePackTraverser to expose more method for this logic. also add javadocs to explain

File: src/test/java/org/influxdb/MessagePackBatchOptionsTest.java [hunk 26]
```diff
@@ -0,0 +1,271 @@
...
+
+/**
+ * Test the InfluxDB API over MessagePack format
+ *
+ * @author hoan.le [at] bonitoo.io
+ *
+ */
+@RunWith(JUnitPlatform.class)
+@EnabledIfEnvironmentVariable(named = "INFLUXDB_VERSION", matches = "1\\.6|1\\.5|1\\.4")
+public class MessagePackBatchOptionsTest extends BatchOptionsTest {
```
[the remark this answers is not in the document]
Comment (lxhoan) [reply src/test/java/org/influxdb/MessagePackBatchOptionsTest.java]: No, it does not. The reason to extends the tests for MessagePack is due to these 2 differences in MessagePack format (in comparison to JSON):
- Time in MessagePack is encoded as a 64 bit Long value (JSON as String)
- Empty array is encoded as nil (JSON keep as empty)

**Comment by lxhoan** [comment]
@majst01 , @fmachado
I fixed/answered your comments:
- undo formatting
- refactor + add javadocs to make MessagePackTraverser more unambiguous
- unit test for MessagePackTraverser
- checking of  version support at querying time
Please take a look

File: pom.xml [hunk 27]
```diff
@@ -73,6 +73,9 @@
           <exclude>docker-compose.yml</exclude>
         </excludes>
       </resource>
+      <resource>
+        <directory>src/test/resources</directory>
```
Comment (majst01) [inline pom.xml]: Why do we need to include test resources in the artifact ?

**Comment by majst01** [comment]
@lxhoan can you please add a meaningful changelog entry, thanks

**Pull request merged**